src/yunohost/certificate.py

```diff
@@ -0,0 +1,751 @@
+    gid = grp.getgrnam(group).gr_gid
+
+    os.chown(path, uid, gid)
+    os.chmod(path, permissions)
+
+
+def _enable_certificate(domain, new_cert_folder) :
+    logger.info("Enabling the certificate for domain %s ...", domain)
+
+    live_link = os.path.join(CERT_FOLDER, domain)
+
+    # If a live link (or folder) already exists
+    if os.path.exists(live_link) :
+        # If it's not a link ... expect if to be a folder
+        if not os.path.islink(live_link):
+            # Backup it and remove it
+            _backup_current_cert(domain)
+            shutil.rmtree(live_link)
+        # Else if it's a link, simply delete it
+        elif os.path.lexists(live_link):
+            os.remove(live_link)
+
+    os.symlink(new_cert_folder, live_link)
+
+    logger.info("Restarting services...")
+
+    for service in ("postfix", "dovecot", "metronome"):
+        _run_service_command("restart", service)
+
+    _run_service_command("reload", "nginx")
+
+
+def _backup_current_cert(domain):
+    logger.info("Backuping existing certificate for domain %s", domain)
+
+    cert_folder_domain = os.path.join(CERT_FOLDER, domain)
+
+    date_tag = datetime.now().strftime("%Y%m%d.%H%M%S")
+    backup_folder = "%s-backups/%s" % (cert_folder_domain, date_tag)
+
+    shutil.copytree(cert_folder_domain, backup_folder)
+
+
+def _check_domain_is_ready_for_ACME(domain):
+    public_ip = yunohost.domain.get_public_ip()
+
+    # Check if IP from DNS matches public IP
+    if not _dns_ip_match_public_ip(public_ip, domain):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_domain_dns_ip_differs_from_public_ip', domain=domain))
+
+    # Check if domain seems to be accessible through HTTP ?
+    if not _domain_is_accessible_through_HTTP(public_ip, domain):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_domain_http_not_working', domain=domain))
+
+
+def _dns_ip_match_public_ip(public_ip, domain):
+    try:
+        resolver = dns.resolver.Resolver()
+        resolver.nameservers = []
+        # FFDN DNS resolvers
+        # See https://www.ffdn.org/wiki/doku.php?id=formations:dns
+        resolver.nameservers.append("80.67.169.12")    # FDN
+        resolver.nameservers.append("80.67.169.40")    #
+        resolver.nameservers.append("89.234.141.66")   # ARN
+        resolver.nameservers.append("141.255.128.100") # Aquilenet
+        resolver.nameservers.append("141.255.128.101") #
+        resolver.nameservers.append("89.234.186.18")   # Grifon
+        resolver.nameservers.append("80.67.188.188")   # LDN
+        answers = resolver.query(domain, "A")
+    except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_error_no_A_record', domain=domain))
+
+    dns_ip = str(answers[0])
+
+    return dns_ip == public_ip
+
+
+def _domain_is_accessible_through_HTTP(ip, domain):
+    try:
+        r = requests.head("http://" + ip, headers={"Host": domain})
```

Comment:
This check doesn't work when an app is installed on domain root and set as public because SSOWat is bypassed and nginx try to redirect to http**S** url.  Using `https` and set `verify=False` works for me.